Allow custom containerd & repository configuration

dcelasun opened the pull request:
This is useful for e.g. adding registry mirrors using Harbor, without
having to modify the default config.

It's also cleaner, with `ccr configure-variables` support.

It also includes a new `ccr update-containerd` command so existing nodes
can be updated without replacing them.

christensenjairus commented on scripts/configure_variables.sh:
```diff
@@ -50,4 +50,10 @@ if [[ "$answer" =~ ^[Yy]$ ]]; then
+CONTAINERD_CONFIG_DIR="$REPO_PATH/scripts/k8s_vm_template/FilesToPlace/containerd.config.d"
+read -p "Create custom containerd config? (y/n): " answer
+if [[ "$answer" =~ ^[Yy]$ ]]; then
+  vim "$CONTAINERD_CONFIG_DIR/custom.toml"
```
Could we change this to not target the custom.toml file, since this could handle multiple files?

dcelasun replied:
What do you mean? By default, `custom.toml` has 2 lines to allow custom registries, and this check allows the user to customize it.

christensenjairus replied:
If we were to create `custom.toml` and allow custom registries for everyone, would it complain if there's no custom registry files to load?

dcelasun replied:
No, that should be fine. If the directory doesn't exist or is empty, this file is a no-op.

christensenjairus replied:
Sweet. Let's leave it static then & not give the user this option.

christensenjairus commented on scripts/k8s_vm_template/FilesToPlace/containerd.certs.d/README.md:
```diff
@@ -0,0 +1,21 @@
+[host."https://my-mirror.example.com"]
+  capabilities = ["pull", "resolve"]
+  [host."https://my-mirror.example.com".header]
+    authorization = ["Basic <BASE64_ENCODED_CREDENTIALS>"]
```
Could you add something to the .gitignore so nobody accidentally commits base64 credentials?

dcelasun replied:
OK, ignored TOML files in this directory since they might all potentially contain credentials.

christensenjairus commented on scripts/k8s_vm_template/FilesToPlace/containerd.certs.d/README.md:
```diff
@@ -0,0 +1,21 @@
+
+As of containerd v1.5, registry configuration is handled with `config_path`. See [documentation](https://github.com/containerd/containerd/blob/main/docs/hosts.md#registry-configuration---introduction).
+
+For each registry, create a `<registry-name>/hosts.toml` file. For example if you have a mirror for Docker Hub, create `docker.io/hosts.toml` with:
```
Do we need multiple files for this, or could we reduce it just to the custom.toml file you mentioned in `configure-variables`?

dcelasun replied:
The old way of doing it in a single file is deprecated and containerd 1.5+ wants one file per registry.

christensenjairus reviewed (commented):
Great addition, I would like to use this myself :) 

Just a few tweaks and we can merge it.

dcelasun commented:
Btw, I changed the branch to `main` to avoid conflicts with your changes on `development`. Let me know if I should change it back and just fix the conflicts :)

christensenjairus commented:
> Btw, I changed the branch to `main` to avoid conflicts with your changes on `development`. Let me know if I should change it back and just fix the conflicts :)

I would like it changed back if you don't mind! 

But I can understand there's a lot of changes in `development` that may make it harder for you to test. Biggest one being that you'll have to reformat your `clusters.tf` file and make a `networks.tf` file if you use Unifi.

I'm curious to know what you think of the changes. I've done my best to validate that everything works as expected, but I've had to change a lot to be able to handle my own complex use case, which required multiple nics & network configurations. I'm open to feedback there.

dcelasun commented:
OK, I'll change it back. I don't use Unifi at all, so let me give this a try over the weekend.

dcelasun committed:
Allow custom containerd & repository configuration

This is useful for e.g. adding registry mirrors using Harbor, without
having to modify the default config.

It's also cleaner, with `ccr configure-variables` support.

It also includes a new `ccr update-containerd` command so existing nodes
can be updated without replacing them.

dcelasun commented:
Rebased on top of `development`. Should be good now :)

dcelasun commented:
Btw, I'm working on a new Ansible playbook so containerd configuration can be updated for existing VMs. Otherwise, any change would require updating the VM template and replacing all VMs :/

dcelasun commented:
All done, added `ccr update-containerd` and updated the PR description. I tested this locally in my cluster where I'm using Harbor as a pull-through cache for major Docker repositories.

Ready for review @christensenjairus :)
